### PR TITLE
Feature/21401 extract adapter sdk

### PR DESCRIPTION
**Pull request by DC2-DanielKrueger**
https://hivemq.kanbanize.com/ctrl_board/57/cards/21401/details/

### CODE REVIEW

**Review by Shadowsph3r3 — APPROVED** [review]

File: hivemq-edge/src/main/java/com/hivemq/protocols/PerSubscriptionSampler.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return CompletableFuture.failedFuture(new InterruptedException());
+        }
+        final PollingOutputImpl pollingOutput = new PollingOutputImpl(new ProtocolAdapterDataSampleImpl());
+        try {
+            perSubscriptionProtocolAdapter.poll(new PollingInputImpl(pollingContext), pollingOutput);
+        }catch (Throwable t){
+            pollingOutput.fail(t);
+            throw t;
+        }
+        final CompletableFuture<PollingOutputImpl.PollingResult> outputFuture = pollingOutput.getOutputFuture();
```
Comment (DC2-DanielKrueger) [inline hivemq-edge/src/main/java/com/hivemq/protocols/PerSubscriptionSampler.java:50]: we need to add a timeout

File: hivemq-edge/src/main/java/com/hivemq/protocols/PerSubscriptionSampler.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        outputFuture.thenAccept(pollingResult -> {
+            if (pollingResult == PollingOutputImpl.PollingResult.SUCCESS) {
+                this.captureDataSample(pollingOutput.getDataSample(), pollingContext);
+            } else if (pollingResult == PollingOutputImpl.PollingResult.FAILURE) {
+                // TODO LOG
```
Comment (schaebo) [inline hivemq-edge/src/main/java/com/hivemq/protocols/PerSubscriptionSampler.java:55]: TODO still there

File: hivemq-edge/src/main/java/com/hivemq/protocols/PollingOutputImpl.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    public void finish() {
+        if(dataSample.getDataPoints().isEmpty()){
+            outputFuture.complete(PollingResult.NO_DATA);
+        } else {
+            outputFuture.complete(PollingResult.SUCCESS);
+        }
+    }
+
+    @Override
+    public void fail(final @NotNull Throwable t) {
```
Comment (schaebo) [inline hivemq-edge/src/main/java/com/hivemq/protocols/PollingOutputImpl.java:48]: A (human readable) error message would be interesting here, too.

Maybe two fail methods.

fail(String errorMsg)
fail(String errorMsg, Throwable cause)

File: modules/hivemq-edge-module-http/src/test/java/com/hivemq/edge/adapters/http/HttpProtocolAdapterTest.java [hunk 4]
```diff
@@ -1,84 +1,119 @@
...
+        final ProtocolAdapterPublishBuilder protocolAdapterPublishBuilder =
                 new ProtocolAdapterPublishBuilderImpl("hivemq", sendCallback);
         protocolAdapterPublishBuilder.withAdapter(httpProtocolAdapter);
         when(publishService.publish()).thenReturn(protocolAdapterPublishBuilder);
+
+        httpProtocolAdapter = new HttpProtocolAdapter(HttpProtocolAdapterInformation.INSTANCE, protocolAdapterInput);
     }
 
+
+    /*
```
Comment (schaebo) [inline modules/hivemq-edge-module-http/src/test/java/com/hivemq/edge/adapters/http/HttpProtocolAdapterTest.java:58]: uncomment or remove ?

File: modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaProtocolAdapterInformation.java [hunk 5]
```diff
@@ -53,8 +55,28 @@ private OpcUaProtocolAdapterInformation() {
...
     }
+
+    @Override
+    public @NotNull String getAuthor() {
+        return "HIVEMQ";
+    }
+
+    @Override
+    public @Nullable ProtocolAdapterCategory getCategory() {
+        return null;
```
Comment (schaebo) [inline modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaProtocolAdapterInformation.java:75]: should return a category. INDUSTRIAL imho.

File: modules/hivemq-edge-module-plc4x/src/main/java/com/hivemq/edge/adapters/plc4x/types/ab/ABProtocolAdapterInformation.java [hunk 6]
```diff
@@ -56,24 +55,39 @@ protected ABProtocolAdapterInformation() {
     }
 
     @Override
-    public String getVersion() {
-        return super.getVersion() + " (BETA)";
+    public @NotNull String getUrl() {
+        return "TODO";
```
Comment (schaebo) [inline modules/hivemq-edge-module-plc4x/src/main/java/com/hivemq/edge/adapters/plc4x/types/ab/ABProtocolAdapterInformation.java:59]: TODO

**Review by schaebo — APPROVED** [review]

File: hivemq-edge/src/main/java/com/hivemq/protocols/PerSubscriptionSampler.java (hunk 2 above)
Comment (DC2-DanielKrueger) [reply hivemq-edge/src/main/java/com/hivemq/protocols/PerSubscriptionSampler.java:55]: yeah that place needed some love and care anyway

File: modules/hivemq-edge-module-plc4x/src/main/java/com/hivemq/edge/adapters/plc4x/types/ab/ABProtocolAdapterInformation.java (hunk 6 above)
Comment (DC2-DanielKrueger) [reply modules/hivemq-edge-module-plc4x/src/main/java/com/hivemq/edge/adapters/plc4x/types/ab/ABProtocolAdapterInformation.java:59]: oh thats from rebasing. That shouldnt be here

**Review by schaebo — APPROVED** [review]

**Pull request merged**